File: 13_Hash Tables/python/hash_tables_basics.py
"""
Hash Tables - Python Implementation
LeetCode Problems: 1, 217, 242, 387, 49, 347, 202, 205, 128
"""

from typing import List, Dict, Set, Tuple
from collections import Counter, defaultdict
import heapq


# ============= TWO SUM =============

def two_sum(nums: List[int], target: int) -> List[int]:
    """
    LeetCode 1: Two Sum
    Find two numbers that add up to target.
    
    Time: O(n) | Space: O(n)
    Pattern: Hash map for complement lookup
    """
    seen = {}
    
    for i, num in enumerate(nums):
        complement = target - num
        if complement in seen:
            return [seen[complement], i]
        seen[num] = i
    
    return []


# ============= CONTAINS DUPLICATE =============

def contains_duplicate(nums: List[int]) -> bool:
    """
    LeetCode 217: Contains Duplicate
    Check if array has duplicate elements.
    
    Time: O(n) | Space: O(n)
    Pattern: Set for O(1) deduplication check
    """
    seen = set()
    
    for num in nums:
        if num in seen:
            return True
        seen.add(num)
    
    return False


# ============= VALID ANAGRAM =============

def is_anagram(s: str, t: str) -> bool:
    """
    LeetCode 242: Valid Anagram
    Check if two strings are anagrams (same chars, different order).
    
    Time: O(n) | Space: O(1) - max 26 lowercase letters
    Pattern: Character frequency counting
    """
    if len(s) != len(t):
        return False
    
    char_count = Counter(s)
    
    for char in t:
        if char not in char_count:
            return False
        char_count[char] -= 1
        if char_count[char] < 0:
            return False
    
    return True


# ============= FIRST UNIQUE CHARACTER =============

def first_unique_char(s: str) -> int:
    """
    LeetCode 387: First Unique Character in a String
    Find first character that appears only once.
    
    Time: O(n) | Space: O(1) - max 26 letters
    Pattern: Frequency count + iterate to find unique
    """
    char_count = Counter(s)
    
    for i, char in enumerate(s):
        if char_count[char] == 1:
            return i
    
    return -1


# ============= GROUP ANAGRAMS =============

def group_anagrams(strs: List[str]) -> List[List[str]]:
    """
    LeetCode 49: Group Anagrams
    Group strings that are anagrams of each other.
    
    Time: O(n*k log k) where k = avg string length | Space: O(n*k)
    Pattern: Use sorted string as grouping key
    """
    anagram_groups = defaultdict(list)
    
    for word in strs:
        # Sort characters to create canonical form
        key = ''.join(sorted(word))
        anagram_groups[key].append(word)
    
    return list(anagram_groups.values())


# ============= TOP K FREQUENT ELEMENTS =============

def top_k_frequent(nums: List[int], k: int) -> List[int]:
    """
    LeetCode 347: Top K Frequent Elements
    Find k most frequent elements.
    
    Time: O(n log k) | Space: O(n)
    Pattern: Frequency count + min heap for top k
    """
    # Count frequencies
    freq_map = Counter(nums)
    
    # Use min heap to keep only top k
    heap = []
    
    for num, freq in freq_map.items():
        heapq.heappush(heap, (freq, num))
        if len(heap) > k:
            heapq.heappop(heap)
    
    return [num for freq, num in heap]


# ============= HAPPY NUMBER =============

def is_happy(n: int) -> bool:
    """
    LeetCode 202: Happy Number
    Determine if number eventually reaches 1 using sum of squares of digits.
    
    Time: O(log n) | Space: O(1)
    Pattern: Cycle detection with hash set
    """
    def get_next(number: int) -> int:
        total_sum = 0
        while number > 0:
            digit = number % 10
            total_sum += digit * digit
            number //= 10
        return total_sum
    
    seen = set()
    
    while n != 1 and n not in seen:
        seen.add(n)
        n = get_next(n)
    
    return n == 1


# ============= ISOMORPHIC STRINGS =============

def is_isomorphic(s: str, t: str) -> bool:
    """
    LeetCode 205: Isomorphic Strings
    Check if strings are isomorphic (character bijective mapping).
    
    Time: O(n) | Space: O(1) - max 26 letters
    Pattern: Character-to-character mapping validation
    """
    if len(s) != len(t):
        return False
    
    s_to_t = {}
    t_to_s = {}
    
    for s_char, t_char in zip(s, t):
        # Check s -> t mapping
        if s_char in s_to_t:
            if s_to_t[s_char] != t_char:
                return False
        else:
            s_to_t[s_char] = t_char
        
        # Check t -> s mapping (ensure bijective)
        if t_char in t_to_s:
            if t_to_s[t_char] != s_char:
                return False
        else:
            t_to_s[t_char] = s_char
    
    return True


# ============= LONGEST CONSECUTIVE =============

def longest_consecutive(nums: List[int]) -> int:
    """
    LeetCode 128: Longest Consecutive Elements
    Find length of longest consecutive elements sequence.
    
    Time: O(n) | Space: O(n)
    Pattern: Set for O(1) lookup + smart iteration (only check sequence starts)
    """
    if not nums:
        return 0
    
    num_set = set(nums)
    longest_streak = 0
    
    for num in num_set:
        # Only start counting from sequence beginnings
        if num - 1 not in num_set:
            current_num = num
            current_streak = 1
            
            while current_num + 1 in num_set:
                current_num += 1
                current_streak += 1
            
            longest_streak = max(longest_streak, current_streak)
    
    return longest_streak


# ============= INTERSECTION OF TWO ARRAYS =============

def intersection(nums1: List[int], nums2: List[int]) -> List[int]:
    """
    LeetCode 349: Intersection of Two Arrays
    Find common elements in two arrays.
    
    Time: O(n + m) | Space: O(min(n, m))
    Pattern: Store smaller array in set, iterate larger
    """
    set1 = set(nums1)
    set2 = set(nums2)
    
    return list(set1 & set2)


# ============= RANSOM NOTE =============

def can_construct(ransom_note: str, magazine: str) -> bool:
    """
    LeetCode 383: Ransom Note
    Check if ransom note can be constructed from magazine letters.
    
    Time: O(n) | Space: O(1) - max 26 letters
    Pattern: Character frequency availability check
    """
    char_count = Counter(magazine)
    
    for char in ransom_note:
        if char not in char_count or char_count[char] == 0:
            return False
        char_count[char] -= 1
    
    return True


# ============= MAJORITY ELEMENT =============

def majority_element(nums: List[int]) -> int:
    """
    LeetCode 169: Majority Element
    Find element appearing more than n/2 times.
    
    Time: O(n) | Space: O(n)
    Pattern: Frequency counting + find max
    """
    count_map = Counter(nums)
    return max(count_map, key=count_map.get)


# ============= TWO POINTER + HASH (FIND PAIRS) =============

def find_pairs(nums: List[int], k: int) -> int:
    """
    Find number of unique pairs with difference k.
    
    Time: O(n) | Space: O(n)
    Pattern: Hash set + two-pass for unique pairs
    """
    if k < 0:
        return 0
    
    num_set = set(nums)
    count = 0
    
    for num in num_set:
        if num + k in num_set:
            count += 1
    
    return count

File: 13_Hash Tables/java/HashTablesBasics.java
/**
 * Hash Tables - Java Implementation
 * LeetCode Problems: 1, 217, 242, 387, 49, 347, 202, 205, 128
 */

import java.util.*;

public class HashTablesBasics {
    
    // ============= TWO SUM =============
    
    /**
     * LeetCode 1: Two Sum
     * Find two numbers that add up to target.
     * 
     * Time: O(n) | Space: O(n)
     * @param nums array of integers
     * @param target sum to find
     * @return indices of two numbers
     */
    public static int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> seen = new HashMap<>();
        
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (seen.containsKey(complement)) {
                return new int[]{seen.get(complement), i};
            }
            seen.put(nums[i], i);
        }
        
        return new int[]{-1, -1};
    }
    
    // ============= CONTAINS DUPLICATE =============
    
    /**
     * LeetCode 217: Contains Duplicate
     * Check if array has duplicate elements.
     * 
     * Time: O(n) | Space: O(n)
     */
    public static boolean containsDuplicate(int[] nums) {
        Set<Integer> seen = new HashSet<>();
        
        for (int num : nums) {
            if (seen.contains(num)) {
                return true;
            }
            seen.add(num);
        }
        
        return false;
    }
    
    // ============= VALID ANAGRAM =============
    
    /**
     * LeetCode 242: Valid Anagram
     * Check if two strings are anagrams.
     * 
     * Time: O(n) | Space: O(1) - max 26 letters
     */
    public static boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        
        int[] charCount = new int[26];
        
        for (int i = 0; i < s.length(); i++) {
            charCount[s.charAt(i) - 'a']++;
            charCount[t.charAt(i) - 'a']--;
        }
        
        for (int count : charCount) {
            if (count != 0) {
                return false;
            }
        }
        
        return true;
    }
    
    // ============= FIRST UNIQUE CHARACTER =============
    
    /**
     * LeetCode 387: First Unique Character in a String
     * Find first character that appears only once.
     * 
     * Time: O(n) | Space: O(1) - max 26 letters
     */
    public static int firstUniqChar(String s) {
        int[] charCount = new int[26];
        
        // Count frequencies
        for (char c : s.toCharArray()) {
            charCount[c - 'a']++;
        }
        
        // Find first unique
        for (int i = 0; i < s.length(); i++) {
            if (charCount[s.charAt(i) - 'a'] == 1) {
                return i;
            }
        }
        
        return -1;
    }
    
    // ============= GROUP ANAGRAMS =============
    
    /**
     * LeetCode 49: Group Anagrams
     * Group strings that are anagrams of each other.
     * 
     * Time: O(n*k log k) | Space: O(n*k)
     */
    public static List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> anagramGroups = new HashMap<>();
        
        for (String word : strs) {
            // Create canonical form (sorted characters)
            char[] chars = word.toCharArray();
            Arrays.sort(chars);
            String key = new String(chars);
            
            anagramGroups.computeIfAbsent(key, k -> new ArrayList<>()).add(word);
        }
        
        return new ArrayList<>(anagramGroups.values());
    }
    
    // ============= TOP K FREQUENT ELEMENTS =============
    
    /**
     * LeetCode 347: Top K Frequent Elements
     * Find k most frequent elements.
     * 
     * Time: O(n log k) | Space: O(n)
     */
    public static int[] topKFrequent(int[] nums, int k) {
        // Count frequencies
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }
        
        // Min heap to keep top k
        PriorityQueue<Integer> heap = new PriorityQueue<>(
            (a, b) -> freqMap.get(a) - freqMap.get(b)
        );
        
        for (int num : freqMap.keySet()) {
            heap.offer(num);
            if (heap.size() > k) {
                heap.poll();
            }
        }
        
        int[] result = new int[k];
        int i = 0;
        while (!heap.isEmpty()) {
            result[i++] = heap.poll();
        }
        
        return result;
    }
    
    // ============= HAPPY NUMBER =============
    
    /**
     * LeetCode 202: Happy Number
     * Check if number eventually reaches 1.
     * 
     * Time: O(log n) | Space: O(1)
     */
    public static boolean isHappy(int n) {
        Set<Integer> seen = new HashSet<>();
        
        while (n != 1 && !seen.contains(n)) {
            seen.add(n);
            n = sumOfSquares(n);
        }
        
        return n == 1;
    }
    
    private static int sumOfSquares(int n) {
        int sum = 0;
        while (n > 0) {
            int digit = n % 10;
            sum += digit * digit;
            n /= 10;
        }
        return sum;
    }
    
    // ============= ISOMORPHIC STRINGS =============
    
    /**
     * LeetCode 205: Isomorphic Strings
     * Check if strings have bijective character mapping.
     * 
     * Time: O(n) | Space: O(1)
     */
    public static boolean isIsomorphic(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        
        Map<Character, Character> sToT = new HashMap<>();
        Map<Character, Character> tToS = new HashMap<>();
        
        for (int i = 0; i < s.length(); i++) {
            char sChar = s.charAt(i);
            char tChar = t.charAt(i);
            
            // Check s -> t mapping
            if (sToT.containsKey(sChar)) {
                if (sToT.get(sChar) != tChar) {
                    return false;
                }
            } else {
                sToT.put(sChar, tChar);
            }
            
            // Check t -> s mapping (bijective)
            if (tToS.containsKey(tChar)) {
                if (tToS.get(tChar) != sChar) {
                    return false;
                }
            } else {
                tToS.put(tChar, sChar);
            }
        }
        
        return true;
    }
    
    // ============= LONGEST CONSECUTIVE =============
    
    /**
     * LeetCode 128: Longest Consecutive Elements
     * Find longest consecutive sequence length.
     * 
     * Time: O(n) | Space: O(n)
     */
    public static int longestConsecutive(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }
        
        Set<Integer> numSet = new HashSet<>();
        for (int num : nums) {
            numSet.add(num);
        }
        
        int longestStreak = 0;
        
        for (int num : numSet) {
            // Only start from sequence beginning
            if (!numSet.contains(num - 1)) {
                int currentNum = num;
                int currentStreak = 1;
                
                while (numSet.contains(currentNum + 1)) {
                    currentNum++;
                    currentStreak++;
                }
                
                longestStreak = Math.max(longestStreak, currentStreak);
            }
        }
        
        return longestStreak;
    }
    
    // ============= INTERSECTION =============
    
    /**
     * LeetCode 349: Intersection of Two Arrays
     * Find common elements in two arrays.
     * 
     * Time: O(n + m) | Space: O(min(n, m))
     */
    public static int[] intersection(int[] nums1, int[] nums2) {
        Set<Integer> set1 = new HashSet<>();
        for (int num : nums1) {
            set1.add(num);
        }
        
        Set<Integer> result = new HashSet<>();
        for (int num : nums2) {
            if (set1.contains(num)) {
                result.add(num);
            }
        }
        
        int[] arr = new int[result.size()];
        int i = 0;
        for (int num : result) {
            arr[i++] = num;
        }
        
        return arr;
    }
    
    // ============= RANSOM NOTE =============
    
    /**
     * LeetCode 383: Ransom Note
     * Check if ransom note can be built from magazine.
     * 
     * Time: O(n) | Space: O(1)
     */
    public static boolean canConstruct(String ransomNote, String magazine) {
        int[] charCount = new int[26];
        
        for (char c : magazine.toCharArray()) {
            charCount[c - 'a']++;
        }
        
        for (char c : ransomNote.toCharArray()) {
            if (charCount[c - 'a'] == 0) {
                return false;
            }
            charCount[c - 'a']--;
        }
        
        return true;
    }
    
    // ============= MAJORITY ELEMENT =============
    
    /**
     * LeetCode 169: Majority Element
     * Find element appearing more than n/2 times.
     * 
     * Time: O(n) | Space: O(n)
     */
    public static int majorityElement(int[] nums) {
        Map<Integer, Integer> countMap = new HashMap<>();
        int majorityCount = nums.length / 2;
        
        for (int num : nums) {
            countMap.put(num, countMap.getOrDefault(num, 0) + 1);
            if (countMap.get(num) > majorityCount) {
                return num;
            }
        }
        
        return -1;
    }
    
    // ============= FIND PAIRS WITH K DIFFERENCE =============
    
    /**
     * Find number of unique pairs with difference k.
     * 
     * Time: O(n) | Space: O(n)
     */
    public static int findPairs(int[] nums, int k) {
        if (k < 0) {
            return 0;
        }
        
        Set<Integer> numSet = new HashSet<>();
        for (int num : nums) {
            numSet.add(num);
        }
        
        int count = 0;
        for (int num : numSet) {
            if (numSet.contains(num + k)) {
                count++;
            }
        }
        
        return count;
    }
}